Opaque ports check

migue opened the pull request:
This is a WIP pull trying to address #6177 

<!--  Thanks for sending a pull request!

If you already have a well-structured git commit message, chances are GitHub
set the title and description of this PR to the git commit message subject and
body, respectively. If so, you may delete these instructions and submit your PR.

If this is your first time, please read our contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md

The title and description of your Pull Request should match the git commit
subject and body, respectively. Git commit messages are structured as follows:

```
Subject

Problem

Solution

Validation

Fixes #[GitHub issue ID]

DCO Sign off
```

Example git commit message:

```
Introduce Pull Request Template

GitHub's community guidelines recommend a pull request template, the repo was
lacking one.

Introduce a `PULL_REQUEST_TEMPLATE.md` file.

Once merged, the
[Community profile checklist](https://github.com/linkerd/linkerd2/community)
should indicate the repo now provides a pull request template.

Fixes #3321

Signed-off-by: Jane Smith <jane.smith@example.com>
```

Note the git commit message subject becomes the pull request title.

For more details around git commits, see the section on Committing in our
contributor guide:
https://github.com/linkerd/linkerd2/blob/main/CONTRIBUTING.md#committing
-->

mateiidavid commented:
Hey @migue thanks for jumping on this so quickly! I saw [your comment](https://github.com/linkerd/linkerd2/issues/6177#issuecomment-850256735) in #6177. 

So far this looks good to me conceptually. If I understand correctly, the output you showed in the comment is just a test (e.g you put the annotation on all of those pods but not the service), right? If so, all is well from my side. One thing I'd suggest is maybe to error out with a smaller message, perhaps only the service name or something similar. 

Look forward to reviewing this in full! 👍🏻

migue commented:
Thanks a lot for taking a look!! 

Yes, I just annotated a bunch of deployments in my Kubernetes dev environment (only to the pods) and run my local build of the CLI.

I will review my set of changes, add the real tests to the source code, and will open a pull for a "formal" review.

Again, thanks a lot for your help!!

migue commented:
I've cleaned the code a bit and added a bunch of test scenarios trying to validate the behavior of the new check. I've tried to follow the same style of the existing code base, but I am totally new here, so, apologies if something does not comply with your existing standards.

Opened it for review.

Thanks a lot!

mateiidavid commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -136,6 +136,10 @@ const (
 	LinkerdCNIPluginChecks CategoryID = "linkerd-cni-plugin"
 
+	// LinkerdOpaquePortsDefinitionChecks adds checks to validate that the opaque-ports
+	// label has been defined both in the service and the corresponding deployment
```
very nitpicky 😅 
```suggestion
	// label has been defined both in the service and the corresponding pods
```

The reason for this nitpicky comment is that the annotation should be put in the pod template spec of the deployment rather than the deployment itself. I've seen many people (myself included) put the annotation directly on the deployment and wondering why it doesn't work. I think it would be helpful for people that are not as familiar with how linkerd works to have this distinction reflected in the comment. Wdyt?

migue replied:
You're totally right!

mateiidavid commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -1415,6 +1419,19 @@ func (hc *HealthChecker) allCategories() []*Category {
+			LinkerdOpaquePortsDefinitionChecks,
+			[]Checker{
+				{
+					description: "the opaque-ports labels have been properly defined",
```
again, very nitpicky:
```suggestion
					description: "the opaque-ports annotations have been properly defined",
```

mateiidavid commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2208,70 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
 	return nil
 }
 
+func (hc *HealthChecker) checkMisconfiguredOpaquePortAnnotations(ctx context.Context, namespace string) error {
```
wdyt about adding a small comment here to describe what misconfiguration means in this case?

migue replied:
I've included a little comment; please, let me know if this makes things clearer?

mateiidavid replied:
Yup! Comments look good 👌🏻 would you mind wrapping them at col 80 please? We try to keep most comments consistent across the codebase in terms of wrapping. Sorry, I should have mentioned it when I put the nitpicky feedback in.

migue replied:
Sure thing!

mateiidavid reviewed (approved):
@migue thank you so much for your contribution! The PR looks really good. I've added a couple of nits. Nothing breaking or irregular that stands out to me so from my pov  it can be 🚢 .

migue commented:
Thanks a lot for taking a look @mateiidavid!

The integrations tests were failing; I've updated the data defined in `test/integration/testdata/check*` but I guess I've probably done something wrong!

Once the tests run again will double-check the outputs and see what I did wrong, sorry.

Migue

mateiidavid commented on test/integration/testdata/check.proxy.golden:
```diff
@@ -63,6 +63,10 @@ linkerd-control-plane-proxy
 
+linkerd-opaque-ports-definition
+-------------------------------
+√ the opaque-ports labels have been properly defined
```
hmmm wonder if this is why you were seeing the failures; labels have been changed for annotations so the test suite might not like it 🤔

migue replied:
I am not really sure, looking at the output data in the CI

Expected:

```
Linkerd core checks
        ===================
        
        kubernetes-api
        --------------
        √ can initialize the client
        √ can query the Kubernetes API
        
        kubernetes-version
        ------------------
        √ is running the minimum Kubernetes API version
        √ is running the minimum kubectl version
        
        linkerd-existence
        -----------------
        √ 'linkerd-config' config map exists
        √ heartbeat ServiceAccount exist
        √ control plane replica sets are ready
        √ no unschedulable pods
        √ control plane pods are ready
        
        linkerd-config
        --------------
        √ control plane Namespace exists
        √ control plane ClusterRoles exist
        √ control plane ClusterRoleBindings exist
        √ control plane ServiceAccounts exist
        √ control plane CustomResourceDefinitions exist
        √ control plane MutatingWebhookConfigurations exist
        √ control plane ValidatingWebhookConfigurations exist
        √ control plane PodSecurityPolicies exist
        
        linkerd-identity
        ----------------
        √ certificate config is valid
        √ trust anchors are using supported crypto algorithm
        √ trust anchors are within their validity period
        √ trust anchors are valid for at least 60 days
        √ issuer cert is using supported crypto algorithm
        √ issuer cert is within its validity period
        √ issuer cert is valid for at least 60 days
        √ issuer cert is issued by the trust anchor
        
        linkerd-webhooks-and-apisvc-tls
        -------------------------------
        √ proxy-injector webhook has valid cert
        √ proxy-injector cert is valid for at least 60 days
        √ sp-validator webhook has valid cert
        √ sp-validator cert is valid for at least 60 days
        
        linkerd-version
        ---------------
        √ can determine the latest version
        √ cli is up-to-date
        
        control-plane-version
        ---------------------
        √ can retrieve the control plane version
        √ control plane is up-to-date
        √ control plane and cli versions match
        
        linkerd-control-plane-proxy
        ---------------------------
        √ control plane proxies are healthy
        √ control plane proxies are up-to-date
        √ control plane proxies and cli versions match
        
        linkerd-opaque-ports-definition
        -------------------------------
        √ the opaque-ports labels have been properly defined
        
        Status check results are √
```

Actual:

```
Linkerd core checks
        ===================
        
        kubernetes-api
        --------------
        √ can initialize the client
        √ can query the Kubernetes API
        
        kubernetes-version
        ------------------
        √ is running the minimum Kubernetes API version
        √ is running the minimum kubectl version
        
        linkerd-existence
        -----------------
        √ 'linkerd-config' config map exists
        √ heartbeat ServiceAccount exist
        √ control plane replica sets are ready
        √ no unschedulable pods
        √ control plane pods are ready
        
        linkerd-config
        --------------
        √ control plane Namespace exists
        √ control plane ClusterRoles exist
        √ control plane ClusterRoleBindings exist
        √ control plane ServiceAccounts exist
        √ control plane CustomResourceDefinitions exist
        √ control plane MutatingWebhookConfigurations exist
        √ control plane ValidatingWebhookConfigurations exist
        √ control plane PodSecurityPolicies exist
        
        linkerd-identity
        ----------------
        √ certificate config is valid
        √ trust anchors are using supported crypto algorithm
        √ trust anchors are within their validity period
        √ trust anchors are valid for at least 60 days
        √ issuer cert is using supported crypto algorithm
        √ issuer cert is within its validity period
        √ issuer cert is valid for at least 60 days
        √ issuer cert is issued by the trust anchor
        
        linkerd-webhooks-and-apisvc-tls
        -------------------------------
        √ proxy-injector webhook has valid cert
        √ proxy-injector cert is valid for at least 60 days
        √ sp-validator webhook has valid cert
        √ sp-validator cert is valid for at least 60 days
        
        linkerd-version
        ---------------
        √ can determine the latest version
        √ cli is up-to-date
        
        control-plane-version
        ---------------------
        √ can retrieve the control plane version
        √ control plane is up-to-date
        √ control plane and cli versions match
        
        linkerd-control-plane-proxy
        ---------------------------
        √ control plane proxies are healthy
        √ control plane proxies are up-to-date
        √ control plane proxies and cli versions match
        
        linkerd-opaque-ports-definition
        -------------------------------
        √ the opaque-ports annotations have been properly defined
        
        Status check results are √
        
        Linkerd extensions checks
        =========================
        
        linkerd-viz
        -----------
        √ linkerd-viz Namespace exists
        √ linkerd-viz ClusterRoles exist
        √ linkerd-viz ClusterRoleBindings exist
        √ tap API server has valid cert
        √ tap API server cert is valid for at least 60 days
        √ tap API service is running
        √ linkerd-viz pods are injected
        √ viz extension pods are running
        √ viz extension proxies are healthy
        ‼ viz extension proxies are up-to-date
            some proxies are not running the current version:
        	* grafana-57487b6558-64bb5 (git-e5a319e5)
        	* tap-injector-59f7489dd9-ssvcm (git-e5a319e5)
        	* web-6885bcbb8f-fxj5j (git-e5a319e5)
        	* tap-674f668557-72hkj (git-e5a319e5)
        	* metrics-api-5ccb85df68-9v8p4 (git-e5a319e5)
            see https://linkerd.io/2/checks/#l5d-viz-proxy-cp-version for hints
        √ viz extension proxies and cli versions match
        √ can initialize the client
        √ viz extension self-check
        
        Status check results are √
```

Not sure where the `linkerd-viz` diff comes from ...

mateiidavid commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -136,6 +136,10 @@ const (
 	/// plugin is installed and ready
 	LinkerdCNIPluginChecks CategoryID = "linkerd-cni-plugin"
 
+	// LinkerdOpaquePortsDefinitionChecks adds checks to validate that the opaque-ports
+	// label has been defined both in the service and the corresponding pods
```
Sorry I'm being so annoying with these comments but can we please wrap them at 80 col? 😅

migue replied:
Sure!

mateiidavid reviewed (commented):
No worries at all! Thanks for being so quick with the changes. We have been experiencing some flakiness with the CI as of late so it might be that there's nothing wrong at all :) I haven't managed to look over the previous run results, will keep an eye out and re-run them for you if the errors are due to flakiness 👍🏻

migue commented:
I think I've addressed all your comments @mateiidavid !! I've updated the integration tests data and see what happens this time!

migue commented:
@mateiidavid I've pushed a new commit; I think I defined the test data in the wrong order so the `String.contains` check was failing :(. This latest commit should fix it 🤞

kleimkuhler reviewed (changes requested):
Thanks for working on this @migue! The changes so far look good, but we're going to need to add a little bit more here.

Right now, the check is only one-way; if a pod has the annotation and its selecting service does not, we issue a warning.

We would also want to do this the other way. If a service has the annotation but it's selected pods do not, a warning should still be issued. While more uncommon, it's possible that some client may send traffic to the service and properly see it as opaque, while another client may send traffic directly to the pod IP. The client that talks directly to the pod IP would never see that the workload is opaque and this could lead to the same errors.

Therefore, the check for misconfigured annotations should assert that the annotation is added to both pods and their selecting service if it is present on either. Let me know if this makes sense or if you have any follow-up questions!

migue commented:
@kleimkuhler  it makes sense; I thought about it but. to be honest, I didn’t know if it would be an issue. Will update the pull accordingly.

Thanks a lot!

mateiidavid commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -3297,6 +3297,224 @@ func TestMinReplicaCheck(t *testing.T) {
+			}
+			err = hc.checkMisconfiguredOpaquePortAnnotations(context.Background(), "test-ns")
+			if err == nil && tc.expected != nil {
+				t.Log("Expected error: nil")
+				t.Logf("Received error: %s\n", err)
+				t.Fatal("test case failed")
```
nit: wdyt about turning this into a single line?

e.g
```suggestion
				t.Fatalf("Expected check to be successful, got: %s", err)
```

Think this would make it a bit more consistent with the rest of the tests. Also, the go test suite will let us know the test has failed so the last line might be slightly redundant.

mateiidavid replied:
nit: wdyt about turning this into a single line?

e.g
```suggestion
				t.Fatalf("Expected check to be successful, got: %s", err)
```

Think this would make it a bit more consistent with the rest of the tests. Also, the go test suite will let us know the test has failed so the last line might be slightly redundant.

mateiidavid commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -3297,6 +3297,224 @@ func TestMinReplicaCheck(t *testing.T) {
+				fmt.Println(tc.expected.Error())
+				fmt.Println(err.Error())
+				if err.Error() != tc.expected.Error() {
+					t.Logf("Expected error: %s\n", tc.expected)
+					t.Logf("Received error: %s\n", err)
+					t.Fatal("test case failed")
```
nit: just as above, wdyt about this as a single line?
```suggestion
					t.Fatalf("Expected error: %s, received: %s", tc.expected, err)
```

also, not sure if we need the new lines, think the Go test runner will insert a newline after each test case so we might be able to skip it.

mateiidavid commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -3297,6 +3297,224 @@ func TestMinReplicaCheck(t *testing.T) {
+				t.Fatal("test case failed")
+			}
+			if err != nil {
+				fmt.Println(tc.expected.Error())
+				fmt.Println(err.Error())
```
hmmm, leftover debugs?

migue replied:
Yeah, sorry, my bad

mateiidavid commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -3297,6 +3297,224 @@ func TestMinReplicaCheck(t *testing.T) {
+		t.Run(fmt.Sprintf("%d", i), func(t *testing.T) {
+			hc.kubeAPI, err = k8s.NewFakeAPI(tc.resources...)
+			if err != nil {
+				t.Fatal(err)
```
nit: wdyt of something like
```suggestion
				t.Fatalf("unexpected error: %s", err)
```

mateiidavid commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2209,90 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+	return false
+}
+
+func missconfiguredOpaqueAnnotation(svcAnnotations map[string]string, podAnnotations map[string]string) bool {
```
veeeeeery nitpicky, there's an extra s in the function name

migue replied:
Done!

mateiidavid commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2209,90 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+func serviceMatchesPod(service corev1.Service, pod corev1.Pod) bool {
+	podSelector := service.Spec.Selector
+	for k, sv := range podSelector {
+		if pv, ok := pod.Labels[k]; ok {
+			if sv != pv {
+				return false
+			}
+		} else {
+			return false
+		}
```
hm, so here it seems that we return after checking just one label? the way I'm reading this: for the labels in the service's selector, if the pod includes that label in its map and if it doesn't match the service we return false. If the pod does not have that label we still return false. should we instead skip this loop iteration? seems that we could return early from this function and miss out on some processing?

migue replied:
Not sure I follow you, sorry. We need to go through all the labels in the service selector to make sure __all__ of them match.

So, if a service has defined a selector like

```
  selector:
    app: app
    tier: backend
```

we need to iterate through all the labels; if we don't find one of them in the pod or find it but the values don't match, we can return early because we know that the service will not match (selectors are combined using AND).

So, if we remove the loop, we wouldn't be checking the full selector, would we?

mateiidavid replied:
Ahhhhh, I see. That makes sense!

mateiidavid replied:
Ahhhhh, I see. That makes sense!

migue commented:
The integration tests were green in the previous execution but they have failed in the current one; I am not sure this failure is related to the current set of changes?

mateiidavid commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2209,90 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+
+	var errStrings []string
+
+	for _, service := range services {
```
one alternative here that would help us get rid of the `serviceMatchesPod()` fn is to use the Kubernetes client and fetch pods for each service based on the service's selector.

 ```suggestion
	for _, service := range services {
	     if service.spec.ClusterIP == "None" {
	       continue // skip headless
	      }
	      
	     labelSet := labels.Set(service.Spec.Selector)
	     pods, err := hc.kubeAPI.CoreV1().Pods(namespace).List(ctx, metav1.ListOptions { LabelSelector: labelSet.AsSelector()}
	     if err != nil {
	         return err
	      }
	     if mistmatch := (....)
	  }
```

This has a few advantages:
1. We can get rid of the pods and error check on lines 2220,2221.
2. We can get rid of the `serviceMatchesPod` function
3. Simplify the logic a little bit.

The downside is we don't have a client-go informer so we would make as many API requests to the Kubernetes server as there are services in that namespace. This might not be a problem though since `check` is meant to be used sparingly, I doubt this would bring anyone's K8s control plane down (or flood it with requests).

wdyt @migue ? I'm not particular to any one way of doing it, think the way it is now works fine -- just an alternative.

migue replied:
That was my first option but wanted to minimize the number of requests to the control plane. If you folks think this is not going to be a problem I think we should go forward with your proposal.

I can just talk for my current setup; we have ~50 services so this would be <50 requests (we have a bunch of headless services) which is not a big deal, but not sure about the number of services in general (I guess people won't have millions of services in a single cluster :D )

mateiidavid replied:
Let's put a hold on this for a bit and see what the second reviewer thinks perhaps. I got bitten by this before so I'm also in two minds about it.

migue replied:
Sounds good!!

Again, thanks a lot reviewing the pull!!

kleimkuhler replied:
So I lean towards @mateiidavid's suggestion on this and going with the selector approach. This could lead to a lot of kube API calls depending on the number of services, but the `check --proxy` command is not in a critical path and we can simplify some of the logic here. If it is a noticeable issue then we can optimize for that down the line.

migue replied:
I am okay with it; as I said, it was my first option but wasn't sure about the potential "performance" implications ...

Will update the pull accordingly

mateiidavid commented on test/integration/testdata/check.cni.golden:
```diff
@@ -77,4 +77,8 @@ linkerd-control-plane-proxy
 
+linkerd-opaque-ports-definition
+-------------------------------
+√ the opaque-ports annotations have been properly defined
```
TIOLI: we could say `opaque-ports are properly annotated` -- might sound a bit more Kubernetes like.

migue replied:
Fine to me; I am far away from being a native speaker so I totally trust you on this :D

Changed!

mateiidavid reviewed (commented):
Thanks for the quick turnaround @migue 🙏🏻  I left some more comments after going through this again; most of them are nits. While not breaking, the nitpicky stuff I commented on would help us keep the codebase a bit more consistent throughout.

I have also left some comments around the `serviceMatchesPod()` function and an alternative way to deal with it.

Thanks again!

migue commented:
@mateiidavid I've pushed all the changes except the one about using the Kube API and perform a request per service

kleimkuhler commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2209,90 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+func misconfiguredOpaquePortAnnotationsInService(service corev1.Service, pods *corev1.PodList) bool {
+	for _, pod := range pods.Items {
+		// The endpoints for headless services are handled differently so don't check this type of services
+		if service.Spec.ClusterIP != "None" && serviceMatchesPod(service, pod) {
```
If the service's cluster IP is `None` then we'll never check any of it's pods, so it'd be good to move this check outside of the `pods` loop.

Additionally if we move forward with the selector change, we'll already know at this point which pods match the service, so the `serviceMatchesPod` check shouldn't be needed.

kleimkuhler commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2209,90 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+	return false
+}
+
+func misconfiguredOpaqueAnnotation(svcAnnotations map[string]string, podAnnotations map[string]string) bool {
```
It may be worth returning an error here instead of a bool. With a bool, we currently don't know where there is a misconfiguration; is it missing on the service or on the pod?

If we returned errors here, then when `svcAnnotation != podAnnotation` we can return an error saying "the pod has the annotation but the service doesn't" or "the service has the annotation but it's pods do not".

If everything is configured correctly, returning `nil` can properly be handled still.

kleimkuhler reviewed (commented):
Thanks for the quick turnaround on this @migue! I do think we should move forward with the selector approach, and left some comments about changes that could improve how we communicate errors back to the user.

A very nit comment I have is could we drop the `-` in `opaque-ports`? We refer to it as "opaque ports" in docs/code, so it's just jumping out to me now as I read it here as "opaque-ports" 😬.

migue commented:
@mateiidavid @kleimkuhler Thanks a lot for taking the time to review the pull.

I've submitted two new commits trying to deal with all your suggestions:

* Using the Kube API to return the lists of pods
* Returning an error instead of a boolean in the helper function
* Using "opaque ports" instead of "opaque-ports"

Let me know what you think now :D

Thanks a lot!

migue commented:
I haven't included changes in the error message we report back to the user and kept the `%s annotation is not properly configured in service %s`. Should we include the "detailed" message we're now returning in the helper function?

My original set of changes included this information but I think it was too verbose, so I am not sure what you would prefer. For me, as a user, I would love to see the detailed message somewhere (the concrete failure)

kleimkuhler commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2210,83 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+	if podAnnotation, found := podAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+		if svcAnnotation, found := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+			if svcAnnotation != podAnnotation {
+				return fmt.Errorf("Both the pod and the service have the annotation but values don't match")
```
To provide more information in these messages, it'd be helpful to provide names so that the user knows where to look! The function will need to accept the service and pods as a whole though so that we can get the name.

`fmt.Errorf("service and pod annotations do not match; service: %s, pod %s",  service.Name, pod.Name)`

And for when one is missing:

`fmt.Errorf("pod has annotation but service does not; service: %s, pod %s", service.Name, pod.Name)`

or

`fmt.Errorf("service has annotation but pod does not; service: %s, pod: %s",  service.Name, pod.Name)`

migue replied:
Tried to improve the messaging here https://github.com/linkerd/linkerd2/pull/6192/commits/5bce6390fa41986277e47a4355e4e3fc929090f4

kleimkuhler reviewed (commented):
Yep now that we have error message that can point to the issue, I think it'd be great to pipe those back to the user! It'd be great to provide names as well.

mateiidavid reviewed (approved):
Looks great! Thanks for being so prompt with changes and for working on this @migue!

📦 🚢

migue commented:
@mateiidavid Thank you all for taking the time to perform the review; I appreciate it! I know it's a tiny contribution but hope this serves as a starting point for future ones :D

kleimkuhler commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2210,84 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+	if podAnnotation, found := podAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+		if svcAnnotation, found := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+			if svcAnnotation != podAnnotation {
+				return fmt.Errorf("Pod %s and service %s have the annotation but values don't match", pod.Name, service.Name)
```
I know these were my suggestions but can we change these to output similar to `kubectl get all` where it prefixes resources with their type? Same comment for the other errors we can return in this function!

```suggestion
				return fmt.Errorf("pod/%s and service/%s have the annotation but values don't match", pod.Name, service.Name)
```

kleimkuhler commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2210,84 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+			continue
+		}
+
+		labelSet := labels.Set(service.Spec.Selector)
```
So here we only want to check a service that has a selector set based off my more general review comment.

kleimkuhler reviewed (commented):
Few small things to go!

The main thing is it looks like we need to only check services that have a selector field set. If not, a service that has no selector seems to match all pods. As an example, inject our demo app emojivoto and annotate `service/voting`

```
$ curl -sL https://run.linkerd.io/emojivoto.yml |linkerd inject - |kubectl apply -f -
...
$ kubectl annotate -n emojivoto pods voting-ff4c54b8d-ckd4r config.linkerd.io/opaque-ports='1234'
...
$ linkerd check
...
linkerd-opaque-ports-definition
-------------------------------
× opaque ports are properly annotated
    Pod voting-ff4c54b8d-ckd4r has the annotation but service kubernetes doesn't
Pod voting-ff4c54b8d-ckd4r has the annotation but service voting-svc doesn't
    see https://linkerd.io/2/checks/#linkerd-opaque-ports-definition for hints
```

Here we can see that because `service/kubernetes` has no selector field, the way we selecting pods will match an empty label set.

Additionally, the formatting needs to be fixed so that the second line is indented the same amount as the first. You can see in the `check` output that the second `Pod voting-ff4c54b8d-ckd4r ...` message is not indented like the first.

kleimkuhler commented:
@migue After some off-GitHub discussion with @adleong I think we should change how we are getting the pods for a service.

In my comment above, we are "fixing" the `service/kubernetes` issue by skipping services that don't have a selector, but this isn't as correct as it could be.

Ideally, in services that don't have a selector then there may be an endpoints object that does point to `targetRef`s with `kind: Pod`; we would still want to check the annotations in this situation!

So, what we should do for getting the pods that belong to a service is to get the service's endpoints object. Then, list through the addresses on that endpoints object and if there is a `targetRef` with `kind: Pod`, we can get the pod name from the object and know we're dealing with a pod that belongs to the service.

As an example, the endpoints object for `service/linkerd-destination` has:

```
subsets:
- addresses:
  - ip: 10.42.0.10
    nodeName: k3d-k3s-default-server-0
    targetRef:
      kind: Pod
      name: linkerd-destination-5b7c65657b-fd7nc
      namespace: linkerd
      resourceVersion: "2617"
      uid: 05c36bfc-3f68-4d5c-afcf-e447eff631bb
  ports:
  - name: grpc
    port: 8086
    protocol: TCP
```

So, if `kind: Pod`, then we can immediately get the full pod name with `targetRef.name` and `targetRef.namespace`.

migue commented:
@kleimkuhler I've included a few more commits trying to deal with your latest proposal (including the little formatting stuff :D ). 

Let me know what you think!

Thanks a lot.

Migue

mateiidavid commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2214,94 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+		pods := make([]*corev1.Pod, 0)
+		for _, subset := range endpoint.Subsets {
+			for _, addr := range subset.Addresses {
+				if addr.TargetRef.Kind == "Pod" {
```
I think we should check if `TargetRef` is nil first otherwise this might result in a nil ptr. dereference. I wrote a [Go playground example](https://play.golang.org/p/aJhoVAjmAzn) to illustrate this.

Basically, since `TargetRef` is a pointer to an object, accessing its attributes might panic if there is no memory address to go to (when the object is nil).

migue replied:
I didn't add the check because I thought you couldn't have an `Address` with no `TargetRef`, could you?

migue replied:
I think I was wrong?

```
// Reference to object providing the endpoint.
// +optional
TargetRef *ObjectReference `json:"targetRef,omitempty" protobuf:"bytes,2,opt,name=targetRef"`
```

Will add the extra check

mateiidavid replied:
It may be possible to have one. This happens typically in scenarios where the Service does not include a selector and the user wants the service to point to an external IP (and/or the Endpoints object is created manually).

For example, with linkerd we have addresses with no refs in multicluster scenarios. We create our own synthetic Endpoints to point to the gateway, this would look like this:

```
apiVersion: v1
kind: Endpoints
subsets:
- addresses:
  - ip: 172.20.0.6
  ports:
  - name: mc-probe
    port: 4191
    protocol: TCP
```

I think it'd be worth having the check in just for good measure in case the namespace may include a mirrored service with the Endpoints object similar to the snippet above. Services without a selector may or may not have the target ref depending on how they were created and how they are managed.

kleimkuhler commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -1428,6 +1427,11 @@ func (hc *HealthChecker) allCategories() []*Category {
 					description: "opaque ports are properly annotated",
 					hintAnchor:  "linkerd-opaque-ports-definition",
 					check: func(ctx context.Context) error {
+						if hc.DataPlaneNamespace == "" {
+							// when checking proxies in all namespaces, this check is a no-op
+							return nil
+						}
```
So right now when running `linkerd check` without specifying a namespace, the check does not issue any warnings anymore.

I think when a namespace is not specified, we should default to all namespaces. That way, when using the example I provided above when I install the emojivoto app and annotate a voting pod but not the voting service, we still see an error issued.

If we remove this check right here, we can actually remove the `namespace` argument from the `checkMisconfiguredOpaquePortAnnotations` function.

This is because in `checkMisconfiguredOpaquePortAnnotations` when we get services with `services, err := hc.GetServices(ctx)`, we have already use the `hc.DataPlaneNamespace` if it is specified. If it is not set, then we default to all namespaces.

So, if we remove this check and remove the `namespace` argument from `checkMisconfiguredOpaquePortAnnotations`, this works as I would expect:

```
$ linkerd check
...
linkerd-opaque-ports-definition
-------------------------------
× opaque ports are properly annotated
    pod/voting-ff4c54b8d-h8pdc has the annotation but service/voting-svc doesn't
    see https://linkerd.io/2/checks/#linkerd-opaque-ports-definition for hints

Status check results are ×

$ linkerd check --emojivoto
...
linkerd-opaque-ports-definition
-------------------------------
× opaque ports are properly annotated
    pod/voting-ff4c54b8d-h8pdc has the annotation but service/voting-svc doesn't
    see https://linkerd.io/2/checks/#linkerd-opaque-ports-definition for hints

Status check results are ×

$ linkerd check --linkerd
...
linkerd-opaque-ports-definition
-------------------------------
√ opaque ports are properly annotated

Status check results are √
```

migue replied:
Hello @kleimkuhler ,

I think this has been my bad, sorry. I introduced this because getting a concrete resource by name requires a valid namespace to be passed, otherwise you get an error similar to

```
an empty namespace may not be set when a resource name is provided
```

when you perform an operation like:

```go
endpoint, err := hc.kubeAPI.CoreV1().Endpoints(hc.DataPlaneNamespace).Get(ctx, service.Name, metav1.GetOptions{})
```

I guess we could perform a list operation with a `LabelSelector` and perform a lookup locally against all the results matching the query. I think the same applies while retrieving a particular pod:

```go
pod, err := hc.kubeAPI.CoreV1().Pods(hc.DataPlaneNamespace).Get(ctx, addr.TargetRef.Name, metav1.GetOptions{})
```

I am not intimately familiar with the Go Kube SDK and, of course, I could be missing something obvious :D

kleimkuhler replied:
@migue all good! Yep so when getting the `endpoint` and `pod` you would just pass the `service.Namespace`. It's similar to how we use `service.Name` for the actual service name.

kleimkuhler reviewed (commented):
Thanks for changing this to check using endpoints! This feels like a much better approach 👍. I left a comment about a change in behavior added by a recent commit that I think is worth changing.

I hope all this feedback is not annoying you! You mentioned above:

>  I know it's a tiny contribution but hope this serves as a starting point for future ones

and I just want to say I do not think this is a tiny contribution! I think from the first edge this is released in users are going to start seeing some immediate feedback and could be _really_ helpful in identifying problems early. We really appreciate you working on this feature.

adleong commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -1415,6 +1420,24 @@ func (hc *HealthChecker) allCategories() []*Category {
+			LinkerdOpaquePortsDefinitionChecks,
+			[]Checker{
+				{
+					description: "opaque ports are properly annotated",
```
Since this is a check which validates data plane configuration (i.e. annotations on pods and services in the data plane) rather than anything about the control plane, I think this check belongs in the LinkerdDataPlaneChecks category rather than in a brand new category.  For example, it is conceptually similar to the "data plane service labels are configured correctly" check.  

This will mean that this check will only be run when the user passes the `--proxy` flag to `linkerd check`.

adleong commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2214,94 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+func misconfiguredOpaquePortAnnotationsInService(service corev1.Service, pods []*corev1.Pod) error {
+	for _, pod := range pods {
+		if err := misconfiguredOpaqueAnnotation(service, pod); err != nil {
+			return err
```
Note that if a service has more than one mismatched pod, this will only report the first mismatched pod from that service (in contrast to if there are multiple misconfigured services, they each get reported).

This is probably the right behavior if the service is missing the annotation because the recommended action is for the user to add the annotation to the service no matter how many pods it has.  On the other hand, if pods are missing the annotation, we probably want to report each pod that needs to be updated.

That all said, we don't have to get all the error reporting perfect right now.  We can always come back and improve the error reporting after this PR has merged.

adleong commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2214,94 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+	if podAnnotation, found := podAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+		if svcAnnotation, found := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+			if svcAnnotation != podAnnotation {
+				return fmt.Errorf("pod/%s and service/%s have the annotation but values don't match", pod.Name, service.Name)
```
probably would be good to include the name of the annotation in the error message to make it more obvious what the problem is.  Something like `service/%s includes pod/%s which has a different value for annotation %s`

adleong replied:
(and similarly for the other error messages returned in this function)

adleong commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2214,94 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+	if svcAnnotation, found := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+		if podAnnotation, found := podAnnotations[k8s.ProxyOpaquePortsAnnotation]; found {
+			if svcAnnotation != podAnnotation {
+				return fmt.Errorf("pod/%s and service/%s have the annotation but values don't match", pod.Name, service.Name)
```
This is redundant with the above, isn't it?

Another potential way to structure this function to eliminate the redundancy would be:

```
svcAnnotation, svcAnnotationOk := svcAnnotations[k8s.ProxyOpaquePortsAnnotation]
podAnnotation, podAnnotationOk := podAnnotations[k8s.ProxyOpaquePortsAnnotation]
if svcAnnotationOk && podAnnotationOk {
// check they match
}
if svcAnnotationOk {
// pod annotation is missing
}
if podAnnotationOk {
// svc annotation is missing
}
```

adleong commented on pkg/healthcheck/healthcheck_test.go:
```diff
@@ -1773,12 +1773,11 @@ func TestDataPlanePodLabels(t *testing.T) {
 				if err == nil {
 					t.Fatal("Expected error, got nothing")
 				}
-				fmt.Println(err.Error())
```
thanks for this cleanup 😁

adleong reviewed (commented):
Thanks for taking this up, @migue, and sorry that I didn't get to reviewing this until now.  It looks great!  I just have a few structural suggestions.

migue commented:
Hey @adleong 

Thanks a lot for the review; appreciate it!! I've submitted a bunch of commits trying to address all your concerns (I've marked as resolved those comments whose I've tried to address)

Let me know what you think!

kleimkuhler commented on pkg/healthcheck/healthcheck.go:
```diff
@@ -2191,6 +2203,87 @@ func checkResources(resourceName string, objects []runtime.Object, expectedNames
+		}
+
+		if mismatch := misconfiguredOpaquePortAnnotationsInService(service, pods); mismatch != nil {
+			errStrings = append(
+				errStrings,
+				mismatch.Error(),
+			)
```
To make formatting similar to other check output we can append errors with the formatting that adds the tab and the `*`:

```suggestion
			errStrings = append(
				errStrings,
				fmt.Sprintf("\t* %s", mismatch.Error())
			)
```

Then below, printing errors is just:

```
return fmt.Errorf(strings.Join(errStrings, "\n"))
```

This should lead to output with formatting that makes separate lines a little more obvious:

```
× opaque ports are properly annotated
    * service/emoji-svc has the annotation config.linkerd.io/opaque-ports but pod/emoji-696d9d8f95-8p94s doesn't
    * pod/voting-67bb58c44c-dgt46 and service/voting-svc have the annotation config.linkerd.io/opaque-ports but values don't match
    see https://linkerd.io/2/checks/#linkerd-opaque-ports-definition for hints
```

kleimkuhler reviewed (commented):
Pretty much good to go! Test file needs to be updated now that it's no longer a separate category and I made one last formatting suggestion.

Functionally, this works great and should be good for an approval after this.

migue commented:
I've submitted two additional commits which try to address the latest suggestions.

Thank you very much everyone for taking the time to thoroughly review it!

adleong reviewed (approved):
Great work!

kleimkuhler reviewed (approved):
@migue This looks great thank you so much for the work you put in on it—super excited to see it in use.
